### ISSUE TITLE
If an app changes, the local debug server goes into a broken state

### ISSUE THREAD

**Issue by chanseokoh** [GoogleCloudPlatform/cloud-tools-for-eclipse#604]
This happens only with a debug run.

Steps to reproduce:

1. Create the hello-world Maven app.
1. Debug As > Debug on Server
1. Make a change to `web.xml` (e.g., add a blank line).
   - In the Servers view, the state of the debug server will change to `[Started, Republish]` right away. After a long while, it will eventually change to `[Started]`.

Now, we can observe a few broken things:

1. The console will show the following error message. This is the same error that a user encountered in #596. Note the message `"Detected file changes:"`

   ```
INFO     2016-08-23 20:11:56,261 module.py:402] [default] Detected file changes:
  /usr/local/google/home/chanseok/runtime-gcloud-eclipse/.metadata/.plugins/org.eclipse.wst.server.core/tmp0/nothing/WEB-INF/web.xml
FATAL ERROR in native method: JDWP No transports initialized, jvmtiError=AGENT_ERROR_TRANSPORT_INIT(197)
ERROR    2016-08-23 20:11:56,660 http_runtime.py:396] bad runtime process port ['']
ERROR: transport error 202: connect failed: Connection refused
ERROR: JDWP Transport dt_socket failed to initialize, TRANSPORT_INIT(510)
JDWP exit error AGENT_ERROR_TRANSPORT_INIT(197): No transports initialized [../../../src/share/back/debugInit.c:750]
ERROR    2016-08-23 20:11:56,671 http_proxy.py:116] bad runtime process port ['']
INFO     2016-08-23 20:11:56,772 module.py:788] default: "GET /_ah/warmup HTTP/1.1" 500 285
   ```
1. In the Debug view, the launch configuration will display `<terminated>`. However, the state in the Servers view is `[Started]`. In fact, the server is still running.
1. The server still serves HTTP requests but is in a broken state. Navigating to `http://localhost:8080` will basically show the same console error.

**Comment by briandealwis** [comment]
@chanseokoh: Does this work if you use _Run As_ instead of _Debug As_?

It looks like `dev_appserver` is terminating and relaunching the underlying JVM with the same JVM flags.  Eclipse doesn't ([yet](https://bugs.eclipse.org/bugs/show_bug.cgi?id=499385)) support accepting multiple connections on a launch.

Would you mind trying the branch at https://github.com/GoogleCloudPlatform/gcloud-eclipse-tools/tree/wip-498 and see if it fixes the debug problem?

**Comment by briandealwis** [comment]
The wip-498 branch will deal with the JDWP/JVMTI errors.

I think the other errors are related to #516

**Comment by chanseokoh** [comment]
@briandealwis This doesn't happen for *Run As*.

I tested the wip-498 branch, but it almost all the time hit the JDWP/JVMTI error at startup. (There have been a couple of occasions it started without the error, but it gets the same error anyway.) I think you should be able to reproduce it.

**Comment by briandealwis** [comment]
Weird!  I tried repeating your steps, @chanseokoh, and saw the same problem with JDWP/JVMTI errors.  I did a clean and relaunched and now it works fine (see below).  If it's working correctly, you should see a _Waiting for vm to connect at port XXXX_ item at the bottom of the launch details in the _Debug_ view:

![screen shot 2016-08-24 at 12 05 36 pm](https://cloud.githubusercontent.com/assets/202851/17938238/1a857f14-69f3-11e6-8217-9afff5864ce1.PNG)

```
INFO     2016-08-24 16:04:02,196 module.py:402] [default] Detected file changes:
  /Users/bsd/Manumitting/Projects/GooglePluginEclipse/runtime-gcloud-CLEAR/.metadata/.plugins/org.eclipse.wst.server.core/tmp0/my-hello-world/WEB-INF
  /Users/bsd/Manumitting/Projects/GooglePluginEclipse/runtime-gcloud-CLEAR/.metadata/.plugins/org.eclipse.wst.server.core/tmp0/my-hello-world/WEB-INF/appengine-web.xml
Aug 24, 2016 4:04:02 PM com.google.appengine.tools.development.SystemPropertiesManager setSystemProperties
INFO: Overwriting system property key 'java.util.logging.config.file', value '/Users/bsd/Manumitting/Projects/GooglePluginEclipse/SDKs/google-cloud-sdk/platform/google_appengine/google/appengine/tools/java/config/sdk/logging.properties' with value 'WEB-INF/logging.properties' from '/Users/bsd/Manumitting/Projects/GooglePluginEclipse/runtime-gcloud-CLEAR/.metadata/.plugins/org.eclipse.wst.server.core/tmp0/my-hello-world/WEB-INF/appengine-web.xml'
Aug 24, 2016 4:04:02 PM com.google.appengine.tools.development.AbstractContainerService configure
WARNING: Null value for containerConfigProperties.get(devappserver.portMappingProvider)
Aug 24, 2016 4:04:02 PM com.google.apphosting.utils.jetty.JettyLogger info
INFO: Logging to JettyLogger(null) via com.google.apphosting.utils.jetty.JettyLogger
Aug 24, 2016 4:04:02 PM com.google.apphosting.utils.jetty.JettyLogger info
INFO: jetty-6.1.x
Aug 24, 2016 4:04:03 PM com.google.apphosting.utils.jetty.JettyLogger info
INFO: Started SelectChannelConnector@localhost:60420
Aug 24, 2016 4:04:03 PM com.google.appengine.tools.development.AbstractModule startup
INFO: Module instance default is running at http://localhost:60420/
Aug 24, 2016 4:04:03 PM com.google.appengine.tools.development.AbstractModule startup
INFO: The admin console is running at http://localhost:60420/_ah/admin
Aug 24, 2016 4:04:03 PM com.google.appengine.tools.development.devappserver2.DevAppServer2Impl doStart
INFO: Dev App Server is now running
Aug 24, 2016 4:04:04 PM com.google.appengine.tools.development.LocalResourceFileServlet doGet
WARNING: No file found for: /_ah/warmup
INFO     2016-08-24 16:04:04,182 module.py:788] default: "GET /_ah/warmup HTTP/1.1" 404 83
```

**Comment by chanseokoh** [comment]
Ah, I didn't import the new bundle in the wip-498 branch. The server seems to work without the error in that branch. Cool.

The only problem, however, is that there is a job `Waiting for vm to connect at port ...` running forever (which I can explicitly stop though) as you can also see in your screenshot.

**Comment by briandealwis** [comment]
> The only problem, however, is that there is a job Waiting for vm to connect at port … running forever (which I can explicitly stop though) as you can also see in your screenshot.

Not a problem… it's the feature ;)  The _Debug_ launch viewer shows all processes and debug targets associated with the launch so that they can be individually terminated. That process is waiting for other JVM connections.  It's a teensy bit of clutter, but I don't think we can do anything about it.

**Comment by chanseokoh** [comment]
I see. Is it then possible / does it make sense to make the job a system job so that a user doesn't see it in the status bar?

![selection_002](https://cloud.githubusercontent.com/assets/10523105/17940741/750d8bec-69fe-11e6-8444-e2efa0ae4ba0.png)

**Comment by briandealwis** [comment]
That's standard Eclipse behaviour, but it I can see that will be annoying in this type of situation.  I'll suggest it on the bug.

**Comment by briandealwis** [comment]
I updated the patch to mark the waiting job as a system job and will argue for the change on the Eclipse bug report.

**Comment by chanseokoh** [comment]
@briandealwis This is fixed by #609, isn't it? No longer reproducible.

**Comment by briandealwis** [comment]
Yes
